src/app/wallets/create-on-chain-address.ts

```diff
@@ -31,9 +33,23 @@ export const createOnChainAddress = async (
   return savedOnChainAddress.address
 }
 
-export const createOnChainAddressByWalletPublicId = async (
-  walletPublicId: WalletPublicId,
-): Promise<OnChainAddress | ApplicationError> => {
+export const createOnChainAddressByWalletPublicId = async ({
+  authorizationService,
+  userId,
+  walletPublicId,
+}: {
+  authorizationService: IAuthorizationService
+  userId: UserId
+  walletPublicId: WalletPublicId
+}): Promise<OnChainAddress | ApplicationError> => {
+  const authResult = await authorizationService.checkPermission({
```

Comment:
shound't this be done in a middleware?

Reply:
Authorization can be middleware but when it becomes very granular / complex its better to have it as part of the domain layer.
With the use cases we want to target (merchant / business accounts with RBAC). I think we should just introduce it into the right layer from the beginning - even though right now it could also be a middleware.

Reply:
From Casbin documentation, the recommend way of doing it is with a middleware: https://casbin.org/docs/en/graphql-middlewares

Reply:
Thats funny because if you follow the referenced link https://graphql.org/learn/authorization/ you will read that the official graphql docs consider it best to delegate authorization to the business logic layer - which is the pattern I'm introducing.

It seems like casbin and graphql docs are not suggesting the same thing - or they are but semantics don't match.

In both cases AFAICT the authorization code happens in the resolver via delegating to a business logic function NOT as a GQL middleware (like it is currently)